### PR TITLE
feat: add kendra instance, vpc, permissioning

### LINKED ISSUE #312: Amazon Kendra support
### Describe the feature

Support Amazon Kendra for data ingestion in L3 constructs

### Use Case

Support additional stores

### Proposed Solution

Add a new L3 construct to support Kendra in an ingestion pipeline

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by MichaelWalker-git**
Fixes #312

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

File: lambda/aws-rag-appsync-stepfn-kendra/input_validation/src/requirements.txt
Comment (krokoko) [inline lambda/aws-rag-appsync-stepfn-kendra/input_validation/src/requirements.txt:1]: let's freeze package versions here

File: lambda/aws-rag-appsync-stepfn-kendra/input_validation/src/lambda.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    response = {
+        'isValid':valid,
+        'files':files_to_process
+    }
+
+    return response
+
+def isvalid_file_format(file_name: str) -> bool:
+    file_format = ['.pdf','.txt','json','.doc','.docx','.xlsx']
```
Comment (krokoko) [inline lambda/aws-rag-appsync-stepfn-kendra/input_validation/src/lambda.py:47]: from https://docs.aws.amazon.com/kendra/latest/dg/index-document-types.html there are other file formats supported by Kendra, is there any reason the pipeline doesn't support them ?

File: lambda/aws-rag-appsync-stepfn-kendra/kendra_job_manager/update_job_status.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import boto3
+import json
+import os
+import re
+import datetime
+from datetime import datetime, timezone
+import uuid
+
+ddb_client = boto3.client('dynamodb')
```
Comment (krokoko) [inline lambda/aws-rag-appsync-stepfn-kendra/kendra_job_manager/update_job_status.py:12]: duplicate line

File: lambda/aws-rag-appsync-stepfn-kendra/kendra_job_manager/update_job_status.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+DOCUMENTS_TABLE = os.environ["DOCUMENTS_TABLE"]
+dynamodb_client = boto3.resource('dynamodb')
+
+def lambda_handler(event, context):
+    table = dynamodb_client.Table(DOCUMENTS_TABLE)
+    table.update_item(
```
Comment (krokoko) [inline lambda/aws-rag-appsync-stepfn-kendra/kendra_job_manager/update_job_status.py:18]: what if this fails ? https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb/client/update_table.html section Exceptions

File: lambda/aws-rag-appsync-stepfn-kendra/kendra_sync/start_sync.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ddb_resource = boto3.resource("dynamodb")
+table = ddb_resource.Table(DOCUMENTS_TABLE)
+
+
+def lambda_handler(event, context):
+    print('event notification: ' + json.dumps(event))
+    created = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
+    resp = kendra_client.start_data_source_sync_job(Id=KENDRA_DATA_SOURCE_INDEX_ID, IndexId=KENDRA_INDEX_ID)
+
+    table.put_item(
```
Comment (krokoko) [inline lambda/aws-rag-appsync-stepfn-kendra/kendra_sync/start_sync.py:32]: same as above, what if this fails ?

File: lambda/aws-rag-appsync-stepfn-kendra/s3_file_uploader/src/requirements.txt
Comment (krokoko) [inline lambda/aws-rag-appsync-stepfn-kendra/s3_file_uploader/src/requirements.txt:1]: same as the other deps file, let's freeze versions

File: src/common/helpers/kendra-helper.ts
Comment (krokoko) [inline src/common/helpers/kendra-helper.ts:1]: a lot happening in this file :) we might break it down in smaller pieces

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-kendra/index.ts [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+  createStepFunctionsExecutionHandlerRole,
+  createSyncRunTable, createUpdateKendraJobStatusFn, createDefaultIsolatedVpcProps, createS3FileUploader,
+} from '../../../common/helpers/kendra-helper';
+import { lambdaMemorySizeLimiter } from '../../../common/helpers/utils';
+import { buildVpc } from '../../../common/helpers/vpc-helper';
+
+/**
+ * The properties for the RagAppsyncStepfnKendraProps class.
+ */
+export interface RagAppsyncStepfnKendraProps {
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-rag-appsync-stepfn-kendra/index.ts:45]: we can review the props, seems like the lambda ones are not used in the construct ?
also, seems like the construct is creating a new bucket, so we cannot provide an existing one ?

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-kendra/architecture.png
Comment (krokoko) [inline src/patterns/gen-ai/aws-rag-appsync-stepfn-kendra/architecture.png:1]: is this up to date ? we can add it to the versioned draw io project

File: lambda/aws-rag-appsync-stepfn-kendra/input_validation/src/lambda.py (hunk 1 above)
Comment (scottschreckengaust) [reply lambda/aws-rag-appsync-stepfn-kendra/input_validation/src/lambda.py:47]: Specifically, `ppt`, `rtf`, `csv`, `md`, `html`, and `xml` are listed.

File: lambda/aws-rag-appsync-stepfn-kendra/input_validation/src/requirements.txt
Comment (scottschreckengaust) [reply lambda/aws-rag-appsync-stepfn-kendra/input_validation/src/requirements.txt:1]: Consider minimum versions

File: .projenrc.ts [hunk 6]
```diff
@@ -57,9 +57,10 @@ const project = new awscdk.AwsCdkConstructLibrary({
     '@mrgrain/jsii-struct-builder',
     'typedoc',
     'typedoc-plugin-markdown',
+    'deepmerge',
   ],
   deps: ['cdk-nag'],
-
+  bundledDeps: ['aws-sdk-mock'],
```
Comment (krokoko) [inline .projenrc.ts:63]: note: check if devdeps is enough

File: dependabot.yml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# Package ecosystems to update and where the package manifests are located.
```
Comment (krokoko) [inline dependabot.yml:1]: generated through projen, can be removed

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-kendra/index.ts [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+
+  /**
+     * @summary Constructs a new instance of the RagAppsyncStepfnKendra class.
+     * @param {cdk.App} scope - represents the scope for all the resources.
+     * @param {string} id - this is a scope-unique id.
+     * @param {RagAppsyncStepfnKendraProps} props - user provided props for the construct.
+     * @since 0.0.0
+     * @access public
+   */
+  constructor(scope: Construct, id: string, props: RagAppsyncStepfnKendraProps) {
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-rag-appsync-stepfn-kendra/index.ts:195]: Can we add support for metrics tracking, similar to:
https://github.com/awslabs/generative-ai-cdk-constructs/blob/main/src/patterns/gen-ai/aws-contentgen-appsync-lambda/index.ts#L174C5-L180C7

File: lambda/aws-rag-appsync-stepfn-kendra/kendra_job_manager/src/update_job_status.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
+# SPDX-License-Identifier: MIT-0
```
Comment (scottschreckengaust) [inline lambda/aws-rag-appsync-stepfn-kendra/kendra_job_manager/src/update_job_status.py:1]: Match license for header

**Review by scottschreckengaust — DISMISSED** [review]

File: lambda/aws-rag-appsync-stepfn-kendra/kendra_sync/src/start_sync.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
```
Comment (scottschreckengaust) [inline lambda/aws-rag-appsync-stepfn-kendra/kendra_sync/src/start_sync.py:2]: License

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-kendra/architecture.png
Comment (scottschreckengaust) [inline src/patterns/gen-ai/aws-rag-appsync-stepfn-kendra/architecture.png:1]: Consider adding the tab to `/main/docs/generative_ai_cdk_constructs.drawio`

**Comment by MichaelWalker-git** [comment]
Screenshot 2024-03-29 at 4 53 11 PM

File: src/patterns/gen-ai/aws-qa-appsync-opensearch/cdk.out/synth.lock [hunk 11]
```diff
@@ -0,0 +1 @@
+44084
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-qa-appsync-opensearch/cdk.out/synth.lock:1]: what is this ?

**Review by krokoko — DISMISSED** [review]

**Review by dineshSajwan — APPROVED** [review]

**Review by krokoko — APPROVED** [review]

**Pull request merged**